src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py

```diff
@@ -0,0 +1,218 @@
+        )
+        self.AgentPoolUpgradeSettings = self.__cmd.get_models(
+            "AgentPoolUpgradeSettings",
+            resource_type=self.resource_type,
+            operation_group="agent_pools",
+        )
+
+
+# pylint: disable=too-many-public-methods
+class AKSAgentPoolContext:
+    """Implement getter functions for all parameters in aks_agentpool_create and aks_agentpool_update.
+    """
+    def __init__(
+        self,
+        cmd: AzCliCommand,
+        raw_parameters: Dict,
+        models: AKSAgentPoolModels,
+        decorator_mode: DecoratorMode,
+    ):
+        if not isinstance(raw_parameters, dict):
+            raise CLIInternalError(
+                "Unexpected raw_parameters object with type '{}'.".format(
+                    type(raw_parameters)
+                )
+            )
+        if not validate_decorator_mode(decorator_mode):
+            raise CLIInternalError(
+                "Unexpected decorator_mode '{}' with type '{}'.".format(
+                    decorator_mode, type(decorator_mode)
+                )
+            )
+        self.cmd = cmd
+        self.raw_param = raw_parameters
+        self.models = models
+        self.decorator_mode = decorator_mode
+        self.intermediates = dict()
+        self.agentpool = None
+
+    def attach_agentpool(self, agentpool: AgentPool) -> None:
+        """Attach the AgentPool object to the context.
+
+        The `agentpool` object is only allowed to be attached once, and attaching again will raise a CLIInternalError.
+
+        :return: None
+        """
+        if self.agentpool is None:
+            self.agentpool = agentpool
+        else:
+            msg = "the same" if self.agentpool == agentpool else "different"
+            raise CLIInternalError(
+                "Attempting to attach the `agentpool` object again, the two objects are {}.".format(
+                    msg
+                )
+            )
+
+    def get_resource_group_name(self) -> str:
+        """Obtain the value of resource_group_name.
+
+        Note: resource_group_name will not be decorated into the `agentpool` object.
+
+        This is a required parameter and its value should be provided by user explicitly.
+
+        :return: string
+        """
+        # read the original value passed by the command
+        resource_group_name = self.raw_param.get("resource_group_name")
+
+        # this parameter does not need dynamic completion
+        # this parameter does not need validation
+        return resource_group_name
+
+    def get_cluster_name(self) -> str:
+        """Obtain the value of cluster_name.
+
+        Note: cluster_name will not be decorated into the `agentpool` object.
+
+        This is a required parameter and its value should be provided by user explicitly.
+
+        :return: string
+        """
+        # read the original value passed by the command
+        cluster_name = self.raw_param.get("cluster_name")
```

Comment:
what if the value none?

Reply:
and also other places.

Reply:
This is a required parameter and could not be `None`. It might be an empty string, but won't pass the [validator](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/_validators.py#L108).